Sprint 15

ArtemDubovitskiy opened the pull request:
Прошу проверить самостоятельную задачу по 15 спринту.

Спасибо!)

ArtemDubovitskiy committed:
Добавлено:
- файл модели CoreData
- добалены 3 Entity
- созданы связи между сущностями модели данных
- на экране создания трекера исправлена слабая ссылка на TrackersViewController

ArtemDubovitskiy committed:
Добавлена инициализация CoreData и метод сохранения контекста в AppDelegate

ArtemDubovitskiy committed:
Добавлено:
- слой работы с данными для абстрагирования приложения от CoreData (для дальнейшего заполнения);
- в классе UIColorMarshalling - алгоритмы сериализации и десериализации цвета
- исправлен размер emoji (c 16 на 14)

ArtemDubovitskiy committed:
Добавлено:
- Категория у делегата создания трекера;
- у делегата создания трекера добавлена проверка категории (есть ли созданная категория или нет);
- удалена тестовая категория в TrackersViewController

ArtemDubovitskiy committed:
Добавлен DaysValueTransformer для хранения расписания дней недели в CoreData

ArtemDubovitskiy committed:
Добавлено:
- слой работы с базой данных;
- исправлено отображение плейсхолдеров при запуске приложения и поиске с добавленными трекерами

So1ovyov commented on Tracker/Domain/TrackerCategoryStore.swift:
```diff
@@ -0,0 +1,119 @@
+    // MARK: - Initializers
+    convenience override init() {
+        let context = (UIApplication.shared.delegate as! AppDelegate).persistentContainer.viewContext
+        try! self.init(context: context)
```
Один из пунктов ТЗ `Нет ошибочных и грубых force-операций (force unwrap, force cast)` давай избавимся от force try и обезопасим проект

So1ovyov commented on Tracker/Domain/TrackerCategoryStore.swift:
```diff
@@ -0,0 +1,119 @@
+    
+    // MARK: - Private Properties
+    private let context: NSManagedObjectContext
+    private var fetchedResultsController: NSFetchedResultsController<TrackerCategoryCoreData>!
```
Здесь также нужно избавиться от `force unwrap`

So1ovyov commented on Tracker/Domain/TrackerRecordStore.swift:
```diff
@@ -0,0 +1,114 @@
+    }
+    // MARK: - Private Properties
+    private let context: NSManagedObjectContext
+    private var fetchedResultsController: NSFetchedResultsController<TrackerRecordCoreData>!
```
Я так понимаю по всем файлам типа `Store` нужно перепроверить и избавиться от force операций, чтобы проект соответствовал ТЗ

ArtemDubovitskiy replied:
Да, делал аналогично примеру проекта emojimixer в учебнике. 
Думал что допустимо на данном этапе, поправлю все force unwrap

So1ovyov replied:
Пример в учебнике подсказал, а ТЗ в спринте подкорректировало решение )

ArtemDubovitskiy replied:
Спасибо, стараюсь сразу избегать принудительных операций, буду внимательней)

So1ovyov commented on Tracker/Domain/TrackerStore.swift:
```diff
@@ -0,0 +1,117 @@
+        }
+        
+        guard let schedule = trackersCoreData.schedule else {
+            throw TrackerStoreError.decodingErrorInvalidSchedule
```
Круто, что реализовал для разных кейсов, разные ошибки

So1ovyov commented on Tracker/Resources/AppDelegate.swift:
```diff
@@ -25,4 +28,32 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
+        let container = NSPersistentContainer(name: "TrackerModel")
+        container.loadPersistentStores(completionHandler: { (storeDescription, error) in
+            if let error = error as NSError? {
+                // TODO: Заменить на реализацию кодом обработку ошибки до отправки на ревью:
```
Кажется не заменил или была какая-то другая задумка ?

ArtemDubovitskiy replied:
На этапе разработки я как понял лучше оставить fatalerror для отладки.
Не убрал коммент перед отправкой

So1ovyov commented on Tracker/Resources/AppDelegate.swift:
```diff
@@ -25,4 +28,32 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
+            } catch {
+                context.rollback() // откат изменений
+                let nserror = error as NSError
+                // TODO: Заменить на реализацию кодом обработку ошибки до отправки на ревью:
```
Круто, что описываешь TODO-шки для следующих итераций

So1ovyov commented on Tracker/Trackers/CreateTracker/CreateTrackerViewController.swift:
```diff
@@ -290,15 +292,25 @@ final class CreateTrackerViewController: UIViewController {
+            do {
+                try trackerCategoryStore.addNewTrackerToCategory(to: selectedCategory, tracker: newTracker)
+            } catch {
+                print("Error create new tracker to category: \(error)")
```
В случае ошибки будет ли какая-то UI реакция ? Например Alert или что-то подобное

ArtemDubovitskiy replied:
на данный момент нет, только в терминале

So1ovyov replied:
Было бы круто подумать над показом ошибки для пользователя, в терминале он ничего не увидит )

ArtemDubovitskiy replied:
Спасибо!

So1ovyov commented on Tracker/Trackers/TrackersViewController.swift:
```diff
@@ -261,36 +292,55 @@ extension TrackersViewController: TrackerCollectionViewCellDelegate {
+            do {
+                try self.trackerRecordStore.addTrackerRecord(trackerRecord)
+            } catch {
+                print("Error add a record to TrackerRecord: \(error)")
```
Вижу что во всем ViewController во всех случаях `catch` просто `print` ошибки. Давай добавим либо реакцию на UI чтобы пользователь понимал что ошибка, либо TODO с описание сделать это в следующей итерации

ArtemDubovitskiy replied:
добавлю TODO) спасибо!

ArtemDubovitskiy committed:
Исправления по замечаниям ревьюера:
- в TrackerStore, TrackerCategotyStore и TrackerRecordStore избавился от грубых force-операций
- добавлен алерт в catch в createviewcontroller(ошибка сохранения трекера) и trackerviewcontroller (ошибка отметки/снятия отметки трекера)

So1ovyov commented on Tracker/Trackers/Alert/ErrorReporting.swift:
```diff
@@ -0,0 +1,18 @@
+import UIKit
+
+final class ErrorReporting {
+    static func showAlert(message: String, controller: UIViewController) {
```
Решение через static быстрое но не такое быстрое по перфомансу
Каждый раз когда ты коде будешь вызывать `ErrorReporting.showAlert` то в памяти будет создаваться объект `ErrorReporting` то есть если в коде ты использовал showAlert 5 раз = 5 объектов в памяти
Но если ты создашь 1 раз константу типа `ErrorReporting` и будешь у нее сколько угодно вызывать в коде метод, объект `ErrorReporting` будет в памяти в одном экземпляре
```
private let errorAlertManager = ErrorAlertManager()

errorAlertManager.showAlert()
errorAlertManager.showAlert()
errorAlertManager.showAlert()
errorAlertManager.showAlert()
```

ArtemDubovitskiy replied:
Понял, спасибо, поправлю)

ArtemDubovitskiy replied:
Спасибо за ревью!

ArtemDubovitskiy committed:
Исправления по ревью.
Добавлены константы для класса ErrorReporting в контроллерах.
